imap_processing/hi/l1c/hi_l1c.py

```diff
@@ -0,0 +1,221 @@
+    """
+    de_dataset = load_cdf(l1b_de_path)
+    sensor_str = de_dataset.attrs["Logical_source"].split("_")[-1].split("-")[0]
+    n_esa_step = de_dataset.esa_step.data.size
+    pset_dataset = allocate_pset_dataset(n_esa_step, sensor_str)
+    # TODO: Stored epoch value needs to be consistent across ENA instruments.
+    #    SPDF says this should be the center of the time bin, but instrument
+    #    teams may disagree.
+    pset_dataset.epoch.data[0] = de_dataset.epoch.data[0]
+
+    return pset_dataset
+
+
+def allocate_pset_dataset(n_esa_steps: int, sensor_str: str):
+    """
+    Allocate an empty xarray.Dataset.
+
+    Parameters
+    ----------
+    n_esa_steps : int
+        Number of Electrostatic Analyzer steps to allocate
+    sensor_str : str
+        '45sensor' or '90sensor'
+
+    Returns
+    -------
+    xarray.Dataset
+        Empty xarray.Dataset ready to be filled with data
+    """
+    cdf_manager = CdfAttributeManager(imap_module_directory / "cdf" / "config")
+    cdf_manager.load_global_attributes("imap_hi_global_cdf_attrs.yaml")
+    cdf_manager.load_variable_attributes("imap_hi_variable_attrs.yaml")
+
+    # preallocate coordinates xr.DataArrays
+    coords = dict()
+    # epoch coordinate has only 1 entry for pointing set
+    attrs = cdf_manager.get_variable_attributes("hi_pset_epoch").copy()
+    dtype = attrs.pop("dtype")
+    coords["epoch"] = xr.DataArray(
+        np.empty(1, dtype=dtype),
+        name="epoch",
+        dims=["epoch"],
+        attrs=attrs,
+    )
+    attrs = cdf_manager.get_variable_attributes("hi_pset_esa_step").copy()
+    dtype = attrs.pop("dtype")
+    coords["esa_step"] = xr.DataArray(
+        np.full(n_esa_steps, attrs["FILLVAL"], dtype=dtype),
+        name="esa_step",
+        dims=["esa_step"],
+        attrs=attrs,
+    )
+    # spin angle bins are 0.1 degree bins for full 360 degree spin
+    attrs = cdf_manager.get_variable_attributes("hi_pset_spin_angle_bin").copy()
+    dtype = attrs.pop("dtype")
+    coords["spin_angle_bin"] = xr.DataArray(
+        np.arange(int(360 / 0.1), dtype=dtype),
+        name="spin_angle_bin",
+        dims=["spin_angle_bin"],
+        attrs=attrs,
+    )
+
+    # Allocate the variables
+    data_vars = dict()
+    # despun_z is a 1x3 unit vector that does not have a DEPEND_1.
+    # Define this dict to override the shape produced in full_dataarray
+    var_shapes = {"despun_z": (1, 3)}
+    for var_name in [
+        "despun_z",
+        "hae_latitude",
+        "hae_longitude",
+        "counts",
+        "exposure_times",
+        "background_rates",
+        "background_rates_uncertainty",
+    ]:
+        data_vars[var_name] = full_dataarray(
+            var_name,
+            cdf_manager.get_variable_attributes(f"hi_pset_{var_name}"),
+            coords,
+            shape=var_shapes.get(var_name, None),
+        )
+
+    # Generate label variables
+    data_vars["esa_step_label"] = xr.DataArray(
```

Comment:
Nice, so these labels do work as data variables? We should make a ticket to update the other instruments later.

Reply:
SKTEditor seemed fine with them being variables rather than coordinates. I wrote up ticket #628 to capture this.